File: src/test/java/ev/projects/GameStartTest.java
package ev.projects;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class GameStartTest {

    @Test
    void factoryDisplaysAreFilled() {
        Center center = new Center();
        Board player1Board = new Board();
        Board player2Board = new Board();
        Board player3Board = new Board();
        Game game = new Game(List.of(new Player(player1Board), new Player(player2Board)), center);
        game.start();

        long blue = Arrays.stream(game.factoryDisplays())
                .mapToLong(fd -> fd.giveTiles(Tile.BLUE))
                .sum();
        int yellow = center.giveTiles(Tile.YELLOW, player2Board);
        int red = center.giveTiles(Tile.RED, player3Board);
        int black = center.giveTiles(Tile.BLACK, player1Board);
        int white = center.giveTiles(Tile.WHITE, player2Board);
        assertEquals(100, blue + yellow + red + black + white + game.bagTiles().size());
        long blueFromBag = game.bagTiles().stream().filter(t -> t == Tile.BLUE).count();
        long yellowFromBag = game.bagTiles().stream().filter(t -> t == Tile.YELLOW).count();
        long redFromBag = game.bagTiles().stream().filter(t -> t == Tile.RED).count();
        long blackFromBag = game.bagTiles().stream().filter(t -> t == Tile.BLACK).count();
        long whiteFromBag = game.bagTiles().stream().filter(t -> t == Tile.WHITE).count();
        assertEquals(20, blue + blueFromBag);
        assertEquals(20, yellow + yellowFromBag);
        assertEquals(20, red + redFromBag);
        assertEquals(20, black + blackFromBag);
        assertEquals(20, white + whiteFromBag);
    }

    @Test
    void onePlayerHasStartingMarker() {
        Player player1 = new PlayerMother().newPlayer();
        Player player2 = new PlayerMother().newPlayer();
        Game game = new Game(List.of(player1, player2), new Center(), 0);

        game.start();

        assertTrue(player1.startsRound() && !player2.startsRound());
    }

    @Test
    void playersStartWith0Points() {
        Player player1 = new PlayerMother().newPlayer();
        Player player2 = new PlayerMother().newPlayer();
        Game game = new Game(List.of(player1, player2));

        game.start();

        assertEquals(0, player1.score());
        assertEquals(0, player2.score());
    }
}

File: python/lib/test/game_start_test.py
import unittest

from azul.board import Board
from azul.center import Center
from azul.game import Game
from azul.tile import Tile
from player_mother import PlayerMother


class TestGameStart(unittest.TestCase):
    def test_factory_displays_are_filled(self):
        center = Center()
        player1_board = Board()
        player2_board = Board()
        player3_board = Board()
        game = Game([PlayerMother().new_player(player1_board), PlayerMother().new_player(player2_board)], center)

        blue = sum(fd.give_tiles(Tile.BLUE) for fd in game.factory_displays)
        yellow = center.give_tiles(Tile.YELLOW, player2_board)
        red = center.give_tiles(Tile.RED, player3_board)
        black = center.give_tiles(Tile.BLACK, player1_board)
        white = center.give_tiles(Tile.WHITE, player2_board)
        self.assertEqual(100, blue + yellow + red + black + white + len(game.bag_tiles()))
        blue_from_bag = sum(1 for t in game.bag_tiles() if t == Tile.BLUE)
        yellow_from_bag = sum(1 for t in game.bag_tiles() if t == Tile.YELLOW)
        red_from_bag = sum(1 for t in game.bag_tiles() if t == Tile.RED)
        black_from_bag = sum(1 for t in game.bag_tiles() if t == Tile.BLACK)
        white_from_bag = sum(1 for t in game.bag_tiles() if t == Tile.WHITE)
        self.assertEqual(20, blue + blue_from_bag)
        self.assertEqual(20, yellow + yellow_from_bag)
        self.assertEqual(20, red + red_from_bag)
        self.assertEqual(20, black + black_from_bag)
        self.assertEqual(20, white + white_from_bag)

    def test_one_player_has_starting_marker(self):
        player1 = PlayerMother().new_player()
        player2 = PlayerMother().new_player()

        Game([player1, player2], Center(), 0)

        self.assertTrue(player1.starts_round and not player2.starts_round)

    def test_players_start_with_0_points(self):
        player1 = PlayerMother().new_player()
        player2 = PlayerMother().new_player()

        Game([player1, player2])

        self.assertEqual(0, player1.score)
        self.assertEqual(0, player2.score)